gapic/src/main/com/google/gapid/views/PipelineView.java

```diff
@@ -184,202 +200,298 @@ protected void onUiThreadError(Loadable.Message error) {
   protected void setPipelines(List<API.Pipeline> pipelines) {
     loading.stopLoading();
     disposeAllChildren(stagesContainer);
-    TabFolder folder = createStandardTabFolder(stagesContainer);
-    createPipelineTabs(folder, pipelines);
+    createPipelineTabs(stagesContainer, pipelines);
     stagesContainer.requestLayout();
   }
 
-  private void createPipelineTabs(TabFolder folder, List<API.Pipeline> pipelines) {
+  private void createPipelineTabs(Composite folder, List<API.Pipeline> pipelines) {
+    HashMap<String, StageUI> stageMap = new HashMap<String, StageUI>();
+
     if (!pipelines.isEmpty()) {
-      for (API.Pipeline result : pipelines) {
-        List<API.Stage> stages = result.getStagesList();
+      RowLayout stripLayout = new RowLayout();
+      stripLayout.fill = true;
+      stripLayout.spacing = 0;
+      stripLayout.wrap = true;
+      Composite stripComposite = withLayoutData( createComposite(folder, stripLayout),
+          new GridData(SWT.FILL, SWT.TOP, true, false));
+      Label separator = withLayoutData( new Label(folder, SWT.SEPARATOR | SWT.HORIZONTAL),
+          new GridData(SWT.FILL, SWT.TOP, true, false));
+
+      StackLayout stageStack = new StackLayout();
+      Composite stageComposite = withLayoutData( createComposite(folder, stageStack),
+          new GridData(SWT.FILL, SWT.FILL, true, true));
+
+      for (int pipeIndex = 0; pipeIndex < pipelines.size(); pipeIndex++) {
+        List<API.Stage> stages = pipelines.get(pipeIndex).getStagesList();
+
+        for (int stageIndex = 0; stageIndex < stages.size(); stageIndex++) {
+          API.Stage stage = stages.get(stageIndex);
+          Button stageButton = new Button(stripComposite, SWT.FLAT);
+          stageMap.put(stage.getDebugName(), new StageUI(stageButton, createStage(stageComposite, stage)));
+
+          stageButton.setText(stage.getDebugName());
+          stageButton.addListener(SWT.Selection, e -> {
+            stageStack.topControl = stageMap.get(stage.getDebugName()).stageComposite;
+            stageComposite.layout();
+            Button selectedButton = stageMap.get(selectedStage).stageButton;
+            if (selectedButton != null) {
+              selectedButton.redraw();
+            }
+            selectedStage = stage.getDebugName();
+            selectedButton = stageButton;
+            selectedButton.redraw();
```

Comment:
With your solution there is no need to check the validity of selectedButton anymore.
It is more likely that selectedStage is not found in stageMap, which would cause get() to return null and a null reference exception. So better use something like:
```
            StageUI stageUI = stageMap.get(selectedStage);
            if (stageUI != null) {
              stageUI.stageButton.redraw();
            }
            selectedStage = stage.getDebugName();
            stageButton.redraw();
```